model/riscv_vmem.sail

```diff
@@ -0,0 +1,851 @@
+function offset_of_va(va : bits(64)) -> bits(64) = {
+  let mask : bits(64) = zero_extend(ones(PAGESIZE_BITS));
+  va & mask
+}
+
+// Valid xlen-wide values containing virtual addrs must have upper
+// bits equal to the MSB of the virtual address.
+// Virtual address widths depend on the virtual memory mode.
+// PRIVATE
+function is_valid_vAddr(struct { va_size_bits, _ } : SV_Params,
+			vAddr			   : bits(64)) -> bool =
+ vAddr == sign_extend(vAddr[va_size_bits - 1 .. 0])
+
+// ****************************************************************
+// PTE (Page Table Entry) in PTN (Page Table Node)
+
+// PTE	    MSBs      PPNs	RSW    BITs
+// Sv32	     -	     31..10    9..8    7..0
+// Sv39	   63..54    53..10    9..8    7..0
+// Sv48	   63..54    53..10    9..8    7..0
+
+// MSBs of PTE are reserved for RV64 extensions.
+// There are no available bits on RV32, so these bits will be zeros on RV32.
+
+// For PTW extensions (non-standard)
+type extPte = bits(64)
+
+// PRIVATE: extract msbs of PTE above the PPN
+function msbs_of_PTE(sv_params : SV_Params, pte : bits(64)) -> bits(64) = {
+  let mask : bits(64) = zero_extend(ones(sv_params.pte_msbs_size_bits));
+  (pte >> sv_params.pte_msbs_lsb_index) & mask
+}
+
+// PRIVATE: extract PPNs of PTE
+function PPNs_of_PTE(sv_params : SV_Params, pte : bits(64)) -> bits(64) = {
+  let mask : bits(64) = zero_extend(ones(sv_params.pte_PPNs_size_bits));
+  (pte >> sv_params.pte_PPNs_lsb_index) & mask
+}
+
+// PRIVATE: 8 LSBs of PTEs in Sv32, Sv39, Sv48 and Sv57
+bitfield PTE_LSBs : bits(8) = {
+  D : 7,    // dirty
+  A : 6,    // accessed
+  G : 5,    // global
+  U : 4,    // User
+  X : 3,    // Execute permission
+  W : 2,    // Write permission
+  R : 1,    // Read permission
+  V : 0	    // Valid
+}
+
+// PRIVATE: check if a PTE is a pointer to next level (non-leaf)
+function pte_is_ptr(pte_lsbs : PTE_LSBs) -> bool = (  pte_lsbs.X() == 0b0)
+						   & (pte_lsbs.W() == 0b0)
+						   & (pte_lsbs.R() == 0b0)
+
+// PRIVATE: check if a PTE is valid
+function pte_is_invalid(pte_lsbs : PTE_LSBs) -> bool = (pte_lsbs.V() == 0b0)
+						       | ((pte_lsbs.W() == 0b1)
+							  & (pte_lsbs.R() == 0b0))
+
+// ----------------
+// Check access permissions in PTE
+
+// For (non-standard) extensions: this function gets the extension-available bits
+// of the PTE in extPte, and the accumulated information of the page-table-walk
+// in ext_ptw. It should return the updated ext_ptw in both success and failure cases.
+
+union PTE_Check = {
+  PTE_Check_Success : ext_ptw,
+  PTE_Check_Failure : (ext_ptw, ext_ptw_fail)
+}
+
+// PRIVATE
+function check_PTE_permission(ac       : AccessType(ext_access_type),
+			      priv     : Privilege,
+			      mxr      : bool,
+			      do_sum   : bool,
+			      pte_lsbs : PTE_LSBs,
+			      ext      : extPte,
+			      ext_ptw  : ext_ptw) -> PTE_Check = {
+  let pte_U = pte_lsbs.U();
+  let pte_R = pte_lsbs.R();
+  let pte_W = pte_lsbs.W();
+  let pte_X = pte_lsbs.X();
+  let b : bool =
```

Comment:
`b` is not a very descriptive name. How about `success`?

Reply:
Ok; fixed.